### ISSUE TITLE
Getting closer to original game audio with Bit-depth reduction and Downsampling

### ISSUE THREAD

**Issue by Bill13579** [PsyCommando/ppmdu_2#9]
Thanks so much for making this! I don't think extracting game audio from the PMD Explorers games would be possible at all without all the work you've done, let alone reimporting back into the game!

I see that it been a while since you've last committed to this repo, and you might've figured out some of this already, but for the past few days I've been trying to replicate the NDS's unique sound with the `sf2` file and the MIDI files exported by ppmdu, and while I don't do much reverse engineering, I did have some experience in things related to audio, in particular I wanted to solve the problem of the rendered wave files always sounding _muffled_, with the actual game sounding so much more crisp in comparison...
I saw in `ppmd_audioutil.txt` that you also had this problem but never got to solve it, and so I wanted to share some of the research I did as well as a proof of concept of how close we can get to explorers with Bit-depth reduction and Downsampling.

1. Bit-depth reduction

_Update:_ Found some new information about the NDS hardware, it appears that the DAC itself was limited to outputting 10-bit audio, which was why although the mixer supported 16-bit, when the sound came out it had to be bitcrushed. Interestingly, it appears that the DSi's hardware supports full 16-bits PCM with a 48 kHz sample rate. https://retrocomputing.stackexchange.com/questions/24952/is-sound-generation-on-the-nintendo-ds-always-clipped-to-10-bits
From some of the research I did it seems that the NDS's sound system, while it theoretically did 16-bit audio, after all the internal processing it actually outputted 10-bit audio. (https://www.copetti.org/writings/consoles/nintendo-ds/, https://www.reddit.com/r/emulation/comments/ru5nld/i_really_love_the_sound_of_the_nintendo_ds/) While I'm not sure why this is, it makes sense as the bit-reduction would cause the noise floor to go up, with some of that noise potentially contributing to the crispness of the Explorers' soundtrack. A before and after comparison is shown below with the Menu soundtrack and the second Temporal Tower soundtrack:

(I'll post a link since Github doesn't support uploading WAV's)
[Just using SF2](https://github.com/Bill13579/nds_sound_render/tree/master/Examples/Default)
[With the Bit-depth reduction](https://github.com/Bill13579/nds_sound_render/tree/master/Examples/WithBitReduction)

As expected, it adds a layer of noise. However, the muffling still isn't solved, something to be expected.

2. Downsampling

After bit-depth reduction the next most popular explanation for the sound difference I was able to find was the sample rate, and the resampling done internally. I was somewhat skeptical at first given the lack of progress I was making, but it was true that when resampling, not doing any interpolation between the samples (which was the widespread explanation for the NDS's `ringing`, which you've also discussed in the aforementioned `ppmd_audioutil.txt` file. However doing the downsampling twice or doing it differently from how the NDS does it could very well also lead to different results. Unlike the bit-depth reduction, this had to be done at the sample level, or, at least at the SF2 player level.

I settled on a proof of concept by modifying `rustysynth`, an SF2 player written in Rustlang, as modifying FluidSynth would've been an undertaking I really didn't want to embark on yet. Extracting the SF2 file exported by `ppmdu_2`, I noticed that the sample rates for each of the samples were indeed different than the NDS's internal sample rate for audio, most at around 22000Hz, while the NDS used around 32000Hz for audio I believe. It's great that `ppmdu_2` did _not_ automatically convert the sample rate to something else, since otherwise this would have never worked, but as it appears that the samples are completely the same as the originals, we are able to directly resample the original files inside `rustysynth` (or, the SF2 player), just like how the NDS does it internally, into again a _set_ sample rate of 32729Hz, and here are the results:

https://github.com/Bill13579/nds_sound_render/tree/master/Examples/WithBitReductionAndDownsampling

#### AAAAAAAHHHHHHHHHHHH!!!!!

Was my reaction because I don't know if it's just me but it sounds almost _exactly_ like what the actual game sounds like, at least in melonDS. I cannot contain this excitement, and while I don't know if it's feasible to support this directly in `ppmdu` as it would involve altering the original samples with post-processing (this setting _reaaaallllyyy_ should be in the SF2 players themselves, as interpolation is something done by the sampler, but I guess no one cares for no interpolation anymore in a world of high-fidelity audio and cubic interpolation?), but anyhow the fact that we can get this close is so amazing to me and a testament to your work.

## Differences that Remain

While this really is already incredibly close to the original PMD: Explorers of Sky soundtrack, there were still small differences that I noticed. I'm still not sure what's causing these, they are small, but the connoisseurs might still take note.

1. The Envelope

The envelope for some patches are more truncated in the SF2 version compared to the original in many cases, I think this is the clearest in Spinda's Cafe's soundtrack:

https://github.com/PsyCommando/ppmdu_2/assets/35489013/b1eacd23-60a4-4c5b-990d-3e64eae66962

https://github.com/PsyCommando/ppmdu_2/assets/35489013/4a9e477b-188d-4045-8eda-cb0a0550dc97

(P.s. figured out how to attach wave file, have to convert it into mp4's first...   eek!)

Each note on the main instrument lasts noticeably longer in the original when compared to the SF2 render. I'm wondering if maybe it's different ways SF2 and Procyon handles envelopes?

2. Sizzling

This would be most noticeable in the Menu theme I think. If you focus in on the background chords, the pad thing that plays the chords, in the original there's a sort of jitter if you listen closely while in the SF2 version it's very clean, digital. This might partially be the envelope problem mentioned earlier, or perhaps it's a quirk with the DAC? It's also weird then that this also appears to happen in melonDS, which would just be using the PC's DAC. I'm leaning towards perhaps some kind of CPU tick based thing (not knowledgeable enough about this, but it appears plausible), or maybe just a simple difference in audio handling between SF2 and Procyon? Or perhaps it could be the LFO problem you mentioned in the text file? Either way, this is a pretty small problem, and the envelope problem is probably bigger and might even just fix this one with it.

https://github.com/PsyCommando/ppmdu_2/assets/35489013/f3c779da-aaab-416a-8c63-d6418370f6d7

https://github.com/PsyCommando/ppmdu_2/assets/35489013/0c94ba9b-8878-4d5e-9ba0-6f00b78f26c4

3. ~~Samples that sound a bit weird?~~ (Issue with rustysynth, doesn't seem to be an issue in fluidsynth)

In summary, we can get exceptionally close to the NDS sound using these two methods, this is the proof of concept I made: https://github.com/Bill13579/nds_sound_render
For now it works well for most tracks, the listed problems excluded, but I do hope that one day we can fully replicate the sound of Procyon, it really seems like a very powerful system.
